scanner/src/main/kotlin/ScanResultsStorage.kt

```diff
@@ -154,27 +154,30 @@ abstract class ScanResultsStorage : PackageBasedScanStorage {
             if (results.isEmpty()) {
                 results
             } else {
-                val scanResults = results.toMutableList()
+                val (matchingProvenance, nonMatchingProvenance) = results.partition { it.provenance.matches(pkg) }
 
-                // Only keep scan results whose provenance information matches the package information.
-                scanResults.retainAll { it.provenance.matches(pkg) }
-                if (scanResults.isEmpty()) {
+                if (matchingProvenance.isEmpty()) {
```

Comment:
If the log output only makes sense if `nonMatchingProvenance` is not empty, this could be more directly expressed by just `nonMatchingProvenance.isNotEmpty()` in line 159.

Reply:
I prefer to keep is as-is, because this better expresses that we want to log if there is no matching provenance (and it avoids a double negation). The fact that the logging shows the non-matches is just a detail.

Reply:
I believe the current if condition makes it trickier to understand, if it's code block may get executed in case `nonMatchingProvenance` is actually empty. This is less obvious, which is why I think we should change the condition.